pkg/schemacompat/schemacompat.go

```diff
@@ -0,0 +1,374 @@
+// So there should never be any case when a schema is considered backward-compatible while in fact it is not.
+//
+// If the narrowExisting argument is true, then the LCD (Lowest-Common-Denominator) between existing schema and the new schema
+// is built (when possible), and returned if no incompatible change was detected (like type change, etc...).
+// If the narrowExisting argument is false, the existing schema is untouched and no LCD schema is calculated.
+//
+// In either case, when no errors are reported, it is ensured that either the existing schema or the calculated LCD
+// is a sub-schema of the new schema.
+func EnsureStructuralSchemaCompatibility(fldPath *field.Path, existing, new *apiextensionsv1.JSONSchemaProps, narrowExisting bool) (lcd *apiextensionsv1.JSONSchemaProps, errors utilerrors.Aggregate) {
+	var newInternal, existingInternal apiextensions.JSONSchemaProps
+	apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(existing, &existingInternal, nil)
+	apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(new, &newInternal, nil)
+	newStrucural, err := schema.NewStructural(&newInternal)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+
+	existingStructural, err := schema.NewStructural(&existingInternal)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+
+	lcdStructural := existingStructural.DeepCopy()
+	errs := lcdForStructural(fldPath, existingStructural, newStrucural, lcdStructural, narrowExisting)
+	if len(errs) > 0 {
+		return nil, errs.ToAggregate()
+	}
+	serialized, err := json.Marshal(lcdStructural.ToGoOpenAPI())
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+	var jsonSchemaProps apiextensionsv1.JSONSchemaProps
+	json.Unmarshal(serialized, &jsonSchemaProps)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+	return &jsonSchemaProps, nil
+}
+
+func checkTypesAreTheSame(fldPath *field.Path, existing, new *schema.Structural) (errorList field.ErrorList) {
+	if new.Type != existing.Type {
+		return field.ErrorList{field.Invalid(fldPath.Child("type"), new.Type, "The type of the should not be changed")}
+	}
+	return nil
+}
+
+func checkUnsupportedValidation(fldPath *field.Path, existing, new interface{}, validationName, typeName string) (errorList field.ErrorList) {
+	if !reflect.ValueOf(existing).IsZero() || !reflect.ValueOf(new).IsZero() {
+		return field.ErrorList{field.Forbidden(fldPath, fmt.Sprintf("The '%s' JSON Schema construct is not supported by the Schema negotiation for type '%s'", validationName, typeName))}
+	}
+	return nil
+}
+
+func checkUnsupportedValidationForNumerics(fldPath *field.Path, existing, new *schema.ValueValidation, typeName string) (errorList field.ErrorList) {
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Not, new.Not, "not", typeName)...)
```

Comment:
`append` is variadic, so you could do something like:

```
errorList = append(errorList,
    checkUnsupportedValidation(fldPath, existing.Not, new.Not, "not", typeName),
    checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "allOf", typeName),
    ...
```

Reply:
Yes, I somehow agree, but these `checkUnsupportedValidation` calls (that return a single `field.Error` you're right) are expected to be replaced (in future contributions) by real validation methods that could very likely return a list of errors (because `oneOf`, `anyOf`, etc ... are themselves deep schemas).
So I'd rather let it like that, and have a consistent structure for this comparison logic, if you agree.